### PR TITLE
[#169913813] Read organizations

### PR DESCRIPTION
**Pull request by alexgpeppe**
This PR aims to create an endpoint to get the organizations related to the user.

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#169913813](https://www.pivotaltracker.com/story/show/169913813): Creazione endopint per lettura organizzazione collegata al delegato

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 1df100313c5f0fa3668144ee80cedbd39f7cee48
</p>

File: src/services/organizationService.ts [hunk 1]
```diff
@@ -515,3 +516,88 @@ export async function getOrganizationInstanceFromDelegateEmail(
...
+          model: User
+        }
+      ]
+    });
+
+    const errorsOrOrganizationEithers = organizationInstances.map(
+      fromOrganizationInstanceToOrganizationObject
+    );
+    return right(
+      errorsOrOrganizationEithers.reduce<ReadonlyArray<Organization> | never>(
```
Comment (gunzip) [inline src/services/organizationService.ts]:
```suggestion
      errorsOrOrganizations.reduce<ReadonlyArray<Organization> | never>(
```

File: src/services/organizationService.ts [hunk 2]
```diff
@@ -515,3 +516,88 @@ export async function getOrganizationInstanceFromDelegateEmail(
...
+        (organizations, errorsOrOrganization) => {
+          if (isLeft(errorsOrOrganization)) {
+            throw new Error("One or more organizations are not a valid object");
```
Comment (gunzip) [inline src/services/organizationService.ts]: do not throw. return `left( readable errors )` or `right(organization array)`.

you'd better use `array.sequence(either)(errorsOrOrganization)` to stop at first error

File: src/types/organization.ts [hunk 3]
```diff
@@ -83,3 +86,46 @@ export function fromPublicAdministrationToFoundAdministration(
...
+  };
+  const users =
+    organizationInstance.users &&
+    organizationInstance.users.map(user => ({
+      email: user.email,
+      family_name: user.familyName,
+      fiscal_code: user.fiscalCode,
+      given_name: user.givenName,
+      role: user.role,
+      work_email: user.workEmail || undefined
```
Comment (gunzip) [inline src/types/organization.ts]: why we need the `|| undefined` part ?

File: src/types/organization.ts [hunk 4]
```diff
@@ -83,3 +86,46 @@ export function fromPublicAdministrationToFoundAdministration(
...
+        href: `/organizations/${organizationInstance.ipaCode}`,
+        rel: "edit"
+      }
+    ],
+    name: organizationInstance.name,
+    pec: organizationInstance.pec,
+    registration_status: organizationInstance.registrationStatus,
+    scope: organizationInstance.scope,
+    users
+  }) as Either<Errors, Organization>; // without casting the transpilation stucks
```
Comment (gunzip) [inline src/types/organization.ts]: remove this cast here, we'll try to resolve the issue together

File: src/controllers/organizationController.ts [hunk 5]
```diff
@@ -150,6 +153,56 @@ export default class OrganizationController {
...
+          const errorOrOrganizations = await getAllRegisteredOrganizations();
+          return errorOrOrganizations.fold<
+            | IResponseErrorInternal
+            | IResponseSuccessJson<{ items: ReadonlyArray<Organization> }>
+          >(handleError, organizations => {
+            return ResponseSuccessJson({
+              items: organizations
+            });
+          });
+        case UserRoleEnum.ORG_MANAGER: // can see only his represented organization
```
Comment (gunzip) [inline src/controllers/organizationController.ts]: we need unit tests for every role here

File: src/controllers/organizationController.ts [hunk 6]
```diff
@@ -150,6 +153,56 @@ export default class OrganizationController {
...
+        return ResponseErrorInternal(
+          "An error occurred while reading from the database."
+        );
+      };
+      switch (user.role) {
+        case UserRoleEnum.ADMIN: // can see all the organizations
+          const errorOrOrganizations = await getAllRegisteredOrganizations();
+          return errorOrOrganizations.fold<
+            | IResponseErrorInternal
+            | IResponseSuccessJson<{ items: ReadonlyArray<Organization> }>
```
Comment (gunzip) [inline src/controllers/organizationController.ts]: this must be the type generated from the API spec (currently: `GetOrganizationResults`)

File: src/controllers/organizationController.ts [hunk 7]
```diff
@@ -293,7 +346,9 @@ export default class OrganizationController {
       this.createSignedDocument(unsignedContractPath, signedContractPath),
       this.createSignedDocument(unsignedMandatePath, signedMandatePath)
     ]);
-    const errorsArray = arrayOfMaybeError.filter(isSome);
+    const errorsArray = arrayOfMaybeError.filter(maybeError =>
```
Comment (gunzip) [inline src/controllers/organizationController.ts]: you can use `const errorsArray = arrayOfMaybeError.filter(isSome);`

File: openapi.yml [hunk 8]
```diff
@@ -86,6 +86,20 @@ paths:
...
+      security:
+        - bearerAuth: []
+      responses:
+        <<: *common-responses
+        200:
+          description: The organizations related to the user
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/GetOrganizationResults"
```
Comment (ioggstream) [inline openapi.yml]: I won't repeat the http method in the schema name, eg `GetOrganization` -> `Organization`

File: openapi.yml [hunk 9]
```diff
@@ -608,6 +626,36 @@ components:
           "scope": "NATIONAL",
           "selected_pec_label": "1"
         }
+    GetOrganizationResults:
```
[the remark this answers is not in the document]
Comment (ioggstream) [reply openapi.yml]: Ok for plurals. Won't use a datatype in the schema name, as that may change.

File: src/types/organization.ts (hunk 3 above)
Comment (alexgpeppe) [reply src/types/organization.ts]: The `workEmail` property is optional in the organization object, so its value must be an email string or `undefined`. The queries always return the stored value from the db, which will be the email string or a `null` value, so in this second case we must use `undefined` in order to create a valid organization object.

File: src/types/organization.ts (hunk 4 above)
Comment (alexgpeppe) [reply src/types/organization.ts]: @gunzip, as you can see, the transpilation in the [last build](https://circleci.com/gh/teamdigitale/io-onboarding-pa-api/1111?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) gets stuck after removing this cast.

**Pull request merged**